src/comm.ts

```diff
@@ -0,0 +1,379 @@
+
+  /** 
+   * The target name for the comm channel.
+   *
+   * Read-only
+   */
+  targetName: string;
+
+  /**
+   * The onClose handler.
+   */
+  onClose: (data?: any) => void;
+
+  /**
+   * The onMsg handler.
+   */
+  onMsg: (data: any) => void;
+
+  /**
+   * Open a comm with optional data.
+   */
+  open(data?: any, metadata?: any): IKernelFuture;
+
+  /**
+   * Send a comm message to the kernel.
+   */
+  send(data: any, metadata?: any, buffers?:(ArrayBuffer | ArrayBufferView)[]): IKernelFuture;
+
+  /**
+   * Close the comm.
+   */
+  close(data?: any, metadata?: any): IKernelFuture;
+}
+
+
+/**
+ * Contents of `comm_info` message.
+ */
+export
+interface ICommInfo {
+  /**
+   * Mapping of comm ids to target names.
+   */
+  comms: Map<string, string>;
+}
+
+
+/*
+ * CommManager for a Kernel.
+ *
+ * http://ipython.org/ipython-doc/dev/development/messaging.html#custom-messages
+ */
+export
+class CommManager {
+
+  /**
+   * Create a CommManager instance.
+   */
+  constructor(kernel: IKernel) {
+    this._kernel = kernel;
+    this._kernel.iopubReceived.connect(this._handleKernelMsg, this);
+  }
+
+  /**
+   * Connect to a comm, or create a new one.
+   *
+   * If a client-side comm already exists, it is returned.
+   */
+  connect(targetName: string, commId?: string): Promise<IComm> {
+    if (commId === void 0) {
+      commId = utils.uuid();
+    }
+    var promise = this._comms.get(commId);
+    if (promise) {
+      return promise;
+    }
+    var comm = new Comm(targetName, commId, this._kernel, () => {
+      this._unregisterComm(comm.commId);
+    });
+    promise = Promise.resolve(comm);
+    this._comms.set(commId, promise);
+    return promise;
+  }
+
+  /**
+   * Register the handler for a "comm_open" message on a given targetName.
+   */
+  registerTarget(targetName: string, cb: (comm: IComm, data: any) => void): void {
+    this._targets.set(targetName, cb);
+  }
+
+  /**
+   * Send a 'comm_info_request', and return the contents of the
+   * 'comm_info_reply'.
+   */
+  commInfo(targetName?: string): Promise<ICommInfo> {
+    var contents = {};
+    if (targetName !== void 0) {
+      contents = { target_name: targetName };
+    }
+    var future = sendCommMessage(this._kernel, 'comm_info_request', contents);
+    return new Promise((resolve, reject) => {
+      future.onReply = (msg) => {
+        resolve(msg.content);
+      }
+    });
+  }
+
+  private _handleKernelMsg(kernel: IKernel, msg: IKernelMessage): void {
+    switch(msg.header.msg_type) {
+       case 'comm_open':
+         this._handleOpen(msg);
+         break;
+       case 'comm_msg':
+         this._handleMsg(msg);
+         break;
+       case 'comm_close':
+         this._handleClose(msg);
+         break;
+    }
+  }
+
+  /**
+   * Handle 'comm_open' kernel message.
+   */  
+  private _handleOpen(msg: IKernelMessage): void {
+    var content = msg.content;
+    var promise = loadTarget(content.target_name, content.target_module, 
```

Comment:
This formatting is kind of hard to follow. I've massaged it a bit:

```typescript
private _handleOpen(msg: IKernelMessage): void {
  var content = msg.content;
  var promise = loadTarget(
    content.target_name, content.target_module, this._targets
  ).then((target: (comm: IComm, data: any) => any) => {
    var comm = new Comm(
      content.target_name, 
      content.comm_id, 
      this._kernel,
      () => { this._unregisterComm(content.comm_id); }
    );
    try {
      var response = target(comm, content.data);
    } catch (e) {
      comm.close();
      this._unregisterComm(comm.commId);
      console.error("Exception opening new comm");
      return Promise.reject(e);
    }
    // Regardless of the target return value, we need to
    // then return the comm
    return Promise.resolve(response).then(() => { return comm; });
  }).catch(() => {
    throw Error('Could not open comm'); 
  });
  this._comms.set(content.comm_id, promise);
}
```

But, I'm still not clear on why you have `Promise.resolve(response).then(...)` at the end, why not just return the comm?